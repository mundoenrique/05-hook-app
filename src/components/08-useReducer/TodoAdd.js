import React from 'react';
import useForm from '../../hooks/useForm';

export default function TodoAdd({ handleAddTodo }) {
	const [{ description }, handleInputChange, resetForm] = useForm({
		description: '',
	});

	const handleSubmit = (e) => {
		e.preventDefault();

		if (description.trim().length <= 1) {
			return;
		}

		const newTodo = {
			id: new Date().getTime(),
			desc: description,
			done: false,
		};

		handleAddTodo(newTodo);
		resetForm();
	};

	return (
		<>
			<h4>agregar TODO</h4>
			<form onSubmit={handleSubmit}>
				<div className="mb-3">
					<label htmlFor="description" className="form-label">
						descripci√≥n de la tarea
					</label>
					<input
						type="text"
						className="form-control"
						name="description"
						id="description"
						autoComplete="off"
						value={description}
						onChange={handleInputChange}
					/>
				</div>
				<div className="d-grid gap-2">
					<button type="submit" className="btn btn-primary">
						Submit
					</button>
				</div>
			</form>
		</>
	);
}
