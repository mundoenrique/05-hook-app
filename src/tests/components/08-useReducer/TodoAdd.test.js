import { shallow } from 'enzyme';
import TodoAdd from '../../../components/08-useReducer/TodoAdd';

describe('Pruebas sobre <TodoAdd />', () => {
	const handleAddTodo = jest.fn();

	const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

	test('debe mostrar el componente <TodoAdd />', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('No debe llamar handleAddTodo', () => {
		const formSubmit = wrapper.find('form').prop('onSubmit');

		formSubmit({ preventDefault() {} });

		expect(handleAddTodo).not.toHaveBeenCalled();
		expect(handleAddTodo).toHaveBeenCalledTimes(0);
	});

	test('Debe llamar handleAddTodo 1', () => {
		const value = 'Aprender React';

		const inputOnchnage = wrapper.find('input').prop('onChange');

		inputOnchnage({
			target: {
				value,
				name: 'description',
			},
		});

		const formSubmit = wrapper.find('form').prop('onSubmit');

		formSubmit({ preventDefault() {} });

		expect(handleAddTodo).toHaveBeenCalled();
	});

	test('Debe llamar handleAddTodo 2', () => {
		const value = 'Aprender React';

		wrapper
			.find('input')
			.simulate('change', { target: { value, name: 'description' } });

		wrapper.find('form').simulate('submit', { preventDefault() {} });

		expect(handleAddTodo).toHaveBeenCalledTimes(1);
		expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
		expect(handleAddTodo).toHaveBeenCalledWith({
			id: expect.any(Number),
			desc: value,
			done: false,
		});
	});

	expect(wrapper.find('input').prop('value')).toBe('');
});
