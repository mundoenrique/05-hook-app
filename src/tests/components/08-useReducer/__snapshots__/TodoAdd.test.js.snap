// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <TodoAdd /> debe mostrar el componente <TodoAdd /> 1`] = `
<Fragment>
  <h4>
    agregar TODO
  </h4>
  <form
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="description"
      >
        descripci√≥n de la tarea
      </label>
      <input
        autoComplete="off"
        className="form-control"
        id="description"
        name="description"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-primary"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</Fragment>
`;
